### PR TITLE
Substraction with borrow

### LINKED ISSUE #133: Implement Big UInt Subtraction with Borrow for `modexp`
**Context:** `modexp.yul`

**Description:** 

```solidity
/// @notice Computes `lhs - (rhs + borrow)`, returning the result along with the new borrow.
/// @dev The result is stored from `resPtr` to `resPtr + (WORD_SIZE * nLimbs)`.
bigUIntSBB(lhsPtr, rhsPtr, borrowPtr, nLimbs, resPtr)
```

`lhsPtr`: The pointer to the MSB of the left operand.
`rhsPtr`: The pointer to the MSB of the right operand.
`borrowPtr`: The pointer to the MSB of the borrow.
`nLimbs`: The number of limbs needed to represent the operands.
`resPtr`: The pointer to where you want the result to be stored.

**Recommendation:**

Follow [this implementation](https://github.com/lambdaclass/lambdaworks/blob/main/math/src/unsigned_integer/element.rs#L795).

### PR DESCRIPTION
**Pull request by fkrause98**
## Changes
- Closes #133.
- This PR adds a substraction with borrow function for BigUInts (i.e. a-b, where a > b).
## How to test 
   Paste the below snippet into a contract you can call, 
    my recommendation is to paste it into the Modexp.yul contract
    and then run this test: `cargo test modexp_tests_98`
```yul
// Test1: Simple 2 limbs
// First Number:
//  - First Limb: 0x0...0
//  - Second Limb: 0x0...F
// Second Number:
//  - First Limb: 0x0...0
//  - Second Limb: 0x0...A
// [0x0...0, 0x0...F]
// - 
// [0x0...0, 0x0...A]
// ------------------
// [0x0...0, 0x0...5]
// Should print:
// 0 (first limb)
// 5 (second limb)

mstore(0x00, 0x0)
mstore(0x20, 0xF)

mstore(0x40, 0x0)
mstore(0x60, 0xA)

let retStart := 0x80
let _, borrow := bigUintSubstractionWithBorrow(0x00, 0x40, 2, retStart)

let res := mload(retStart)
let res2 := mload(add(retStart, 0x20))
let res3 := mload(add(retStart, 0x40))
let res4 := mload(add(retStart, 0x60))

console_log(res)
console_log(res2)

// Test2: Borrow works as expected
// First Number:
//  - First Limb: 0xFF
//  - Second Limb: 0xAA
// Second Number:
//  - First Limb: 0xAA
//  - Second Limb: 0xFF
// Should print:
// 54 (first limb)
// ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffab (second limb)



mstore(0x00, 0xFF)
mstore(0x20, 0xAA)

mstore(0x40, 0xAA)
mstore(0x60, 0xFF)


let retStart := 0x80
let _, borrow := bigUintSubstractionWithBorrow(0x00, 0x40, 2, retStart)

let res := mload(retStart)
let res2 := mload(add(retStart, 0x20))
let res3 := mload(add(retStart, 0x40))
let res4 := mload(add(retStart, 0x60))

console_log(res)
console_log(res2)

// Test3: Big number 2 limb substraction
// I've tested this againts our reference (Lambdaworks) and gives the same result.
// First Number:
mstore(0x20, 0xffffaaaaaaaaaaaaaaaaffffffffffffffffaaaaaaaaaaaaaaaa)
mstore(0x40, 0xffffffffffffffffaaaaaaaaaaaaaaaaffffffffffffffffaaaaaaaaaaaaaaaa)
// Second Number:
mstore(0x60, 0xaaaaffffffffffffffffaaaaaaaaaaaaaaaaffffffffffffffff)
mstore(0x80, 0xaaaaaaaaaaaaaaaaffffffffffffffffaaaaaaaaaaaaaaaaffffffffffffffff)

let retStart := 0x100
let _, borrow := bigUintSubstractionWithBorrow(0x20, 0x60, 2, retStart)

// Should print:
// 5554aaaaaaaaaaaaaaab5555555555555554aaaaaaaaaaaaaaab
// 5555555555555554aaaaaaaaaaaaaaab5555555555555554aaaaaaaaaaaaaaab            

let res := mload(retStart)
let res2 := mload(add(retStart, 0x20))

console_log(res)
console_log(res2)
```

### CODE REVIEW

File: precompiles/Modexp.yul [hunk 1]
```diff
@@ -28,6 +28,86 @@ object "ModExp" {
...
+
+            let retStart := 0x100
+            let _, borrow := bigUintSubstractionWithBorrow(0x00, 0x80, 4, retStart)
+
+            let res := mload(retStart)
+            let res2 := mload(add(retStart, 0x20))
+            let res3 := mload(add(retStart, 0x40))
+            let res4 := mload(add(retStart, 0x60))
+
+            console_log(res)
```
Comment (jpcenteno) [inline precompiles/Modexp.yul:106]: Calls to console_log should be removed

**Review by jpcenteno — CHANGES_REQUESTED** [review]

File: precompiles/Modexp.yul [hunk 2]
```diff
@@ -28,6 +28,86 @@ object "ModExp" {
...
+
+            function subLimbsWithBorrow(leftLimb, rightLimb, limbBorrow) -> substractionResult, returnBorrow {
+                let rightPlusBorrow := add(rightLimb, limbBorrow)
+                substractionResult := sub(leftLimb, rightPlusBorrow)
+                if gt(substractionResult, leftLimb) {
+                 returnBorrow := 1
+                }
+            }
+
+            function bigUintSubstractionWithBorrow(lhsPointer, rhsPointer, numberOfLimbs, resultPointer) -> resultPointer, borrow {
```
Comment (jpcenteno) [inline precompiles/Modexp.yul:61]: It is spelled `subtraction`, not `subStraction`

**Review by jpcenteno — CHANGES_REQUESTED** [review]

File: precompiles/Modexp.yul [hunk 3]
```diff
@@ -28,6 +28,100 @@ object "ModExp" {
...
+            //  - Fourth Limb: 0xAAAA
+            // Second Number:
+            //  - First Limb: 0xAAAA
+            //  - Second Limb: 0xFFFF
+            //  - Third Limb: 0xAAAA
+            //  - Fourth Limb: 0xFFF
+            //  
+            // Should print:
+            // 0x5554 (first limb)
+            // 0xffffffffffffaaab (second limb)
```
Comment (jpcenteno) [inline precompiles/Modexp.yul:96]: I wasn't getting the same value as the comments suggested, so I made a python script to test it out:

```python
def bigint_from_limbs(a, b, c, d):
    x = a
    x = (x << 256) | b
    x = (x << 256) | c
    x = (x << 256) | d
    return x

def limbs_from_bigint(bi):
    MASK = (1 << 256) - 1
    d = (bi >> (256 * 0)) & MASK
    c = (bi >> (256 * 1)) & MASK
    b = (bi >> (256 * 2)) & MASK
    a = (bi >> (256 * 3)) & MASK
    return (a, b, c, d)

def print_limbs(title, limbs):
    (a, b, c, d) = limbs
    print(f"{title}: [{hex(a)}, {hex(b)}, {hex(c)}, {hex(d)}]")

lhs_limbs = (0xFFFF, 0xAAAA, 0xFFFF, 0xAAAA)
rhs_limbs = (0xAAAA, 0xFFFF, 0xAAAA, 0xFFF)
print_limbs("lhs_limbs", lhs_limbs)
print_limbs("rhs_limbs", rhs_limbs)

lhs = bigint_from_limbs(*lhs_limbs)
rhs = bigint_from_limbs(*rhs_limbs)
assert lhs_limbs == limbs_from_bigint(lhs)
assert rhs_limbs == limbs_from_bigint(rhs)
print("lhs: ", lhs)
print("rhs: ", rhs)

assert lhs > rhs

res = lhs - rhs
print("res: ", res)
res_limbs = limbs_from_bigint(res)
print_limbs("res_limbs", res_limbs)
```

Which gave me the following result:
```
lhs_limbs: [0xffff, 0xaaaa, 0xffff, 0xaaaa]
rhs_limbs: [0xaaaa, 0xffff, 0xaaaa, 0xfff]
lhs:  101744273178926960064988370776390104958760025131742712410997775312037249184302295032822575714652627321637385044653794456059156179857737103542777570440297726813614998909130177654716915530113012821233856222054627051083490363276862118996650
rhs:  67829515452617973376658913850926736639173350087828474940665183541358166122868684844488766469824485205385066374283620948616688854627385354361415216852260078931923727512825878445441091596645481527124071237319796322089793313955471257767935
res:  33914757726308986688329456925463368319586675043914237470332591770679083061433610188333809244828142116252318670370173507442467325230351749181362353588037647881691271396304299209275823933467531294109784984734830728993697049321390861228715
res_limbs: [0x5554, 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffaaab, 0x5555, 0x9aab]
```

**Review by jpcenteno — CHANGES_REQUESTED** [review]

File: precompiles/Modexp.yul (hunk 1 above)
Comment (jpcenteno) [reply precompiles/Modexp.yul]: I fixed this @ [3643287](https://github.com/lambdaclass/zksync_era_precompiles/pull/149/commits/3643287ecfa28e4e12a898578cf4d48542e7dc6e)

File: precompiles/Modexp.yul (hunk 2 above)
Comment (jpcenteno) [reply precompiles/Modexp.yul]: d886cce

**Comment by jpcenteno** [comment]
I wrote these tests to check this PR:

First, paste this snippet to `precompiles/Playground.yul`:

```yul
object "Playground" {
	code { }
	object "Playground_deployed" {
		code {
            function console_log(val) -> {
                let log_address := 0x000000000000000000636F6e736F6c652e6c6f67
                // A big memory address to store the function selector.
                let freeMemPointer := 0x600
                // store the function selector of log(uint256) in memory
                mstore(freeMemPointer, 0xf82c50f1) // mem[0] = 0xf8...
                // store the first argument of log(uint256) in the next memory slot
                mstore(add(freeMemPointer, 0x20), val)
                // call the console.log contract
                if iszero(staticcall(gas(),log_address,add(freeMemPointer, 28),add(freeMemPointer, 0x40),0x00,0x00)) {
                    revert(0,0)
                }
            }

            // 🤌 PASTE EVERY FUNCTION FROM MODEXP HERE 🤌

            // TESTS

            // lhs = [0]
            // rhs = [0]
            // expected = [0]
            let lhsptr := 0x00
            let rhsptr := 0x20
            let resptr := 0x40
            let numberoflimbs := 1
            mstore(lhsptr, 0x00)
            mstore(rhsptr, 0x00)
            bigUintSubtractionWithBorrow(lhsptr, rhsptr, numberoflimbs, resptr)
            console_log(mload(resptr)) // Should print 0
            

            // lhs = [0xFF]
            // rhs = [0]
            // expected = [0xFF]
            let lhsptr := 0x00
            let rhsptr := 0x20
            let resptr := 0x40
            let numberoflimbs := 1
            mstore(lhsptr, 0xFF)
            mstore(rhsptr, 0x00)
            bigUintSubtractionWithBorrow(lhsptr, rhsptr, numberoflimbs, resptr)
            console_log(sub(mload(resptr), 0xFF) // Should print 0

            // lhs = [0xFF]
            // rhs = [0xF0]
            // expected = [0x0F]
            let lhsptr := 0x00
            let rhsptr := 0x20
            let resptr := 0x40
            let numberoflimbs := 1
            mstore(lhsptr, 0xFF)
            mstore(rhsptr, 0xF0)
            bigUintSubtractionWithBorrow(lhsptr, rhsptr, numberoflimbs, resptr)
            console_log(sub(mload(resptr), 0x0F) // Should print 0

            // 2-limb subtraction without borrow
            // lhs      = [0xAB, 0xCD]
            // rhs      = [0xA0, 0x0D]
            // expected = [0x0B, 0xC0]
            let lhsptr := 0x00
            let rhsptr := 0x40
            let resptr := 0x80
            let numberoflimbs := 2
            mstore(add(lhsptr, 0x00), 0xAB)
            mstore(add(lhsptr, 0x20), 0xCD)
            mstore(add(rhsptr, 0x00), 0xA0)
            mstore(add(rhsptr, 0x20), 0x0D)
            bigUintSubtractionWithBorrow(lhsptr, rhsptr, numberoflimbs, resptr)
            console_log(sub(mload(add(resptr, 0x00)), 0x0B) // Should print 0
            console_log(sub(mload(add(resptr, 0x20)), 0xC0) // Should print 0

            // 2-limb subtraction with borrow
            // lhs      = [0x01, 0x00]
            // rhs      = [0x00, 0x01]
            // expected = [0x00, 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff]
            let lhsptr := 0x00
            let rhsptr := 0x40
            let resptr := 0x80
            let numberoflimbs := 2
            mstore(add(lhsptr, 0x00), 0x01)
            mstore(add(lhsptr, 0x20), 0x00)
            mstore(add(rhsptr, 0x00), 0x00)
            mstore(add(rhsptr, 0x20), 0x01)
            bigUintSubtractionWithBorrow(lhsptr, rhsptr, numberoflimbs, resptr)
            console_log(sub(mload(add(resptr, 0x00)), 0x0) // should print 0
            console_log(sub(mload(add(resptr, 0x20)), 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff) // should print 0

            // 4-limb subtraction with borrow
            // lhs      = [0xFFFF, 0xAAAA, 0xFFFF, 0xAAAA]
            // rhs      = [0xAAAA, 0xFFFF, 0xAAAA, 0xFFF]
            // expected = [0x5554, 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffaaab, 0x5555, 0x9aab]
            let lhsptr := 0x00
            let rhsptr := 0x80
            let resptr := 0x100
            let numberoflimbs := 4

            mstore(add(lhsptr, 0x00), 0xFFFF) // lhs[0] = 0xFFFF
            mstore(add(lhsptr, 0x20), 0xAAAA) // lhs[1] = 0xAAAA
            mstore(add(lhsptr, 0x40), 0xFFFF) // lhs[2] = 0xFFFF
            mstore(add(lhsptr, 0x60), 0xAAAA) // lhs[3] = 0xAAAA

            mstore(add(rhsptr, 0x00), 0xAAAA) // rhs[0] = 0xAAAA
            mstore(add(rhsptr, 0x20), 0xFFFF) // rhs[1] = 0xFFFF
            mstore(add(rhsptr, 0x40), 0xAAAA) // rhs[2] = 0xAAAA
            mstore(add(rhsptr, 0x60), 0xFFF)  // rhs[3] = 0xFFF

            bigUintSubtractionWithBorrow(lhsptr, rhsptr, numberoflimbs, resptr)
            console_log(sub(mload(add(resptr, 0x00)), 0x5554) // res[0], should print 0
            console_log(sub(mload(add(resptr, 0x20)), 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffaaab) // res[1], should print 0
            console_log(sub(mload(add(resptr, 0x40)), 0x5555) // res[2], should print 0
            console_log(sub(mload(add(resptr, 0x60)), 0x9aab) // res[3], should print 0
            

        }
	}
}
```

Then, copy all your functions from `precompiles/ModExp.yul`.

after that, compile the contracts and spin up the `era-test-node` using `make run`.

When the server is ready, run the tests using
```sh
 zksync-era-cli --host localhost --port 8011 call --contract 0x000000000000000000000000000000000000FFFF --function "" --data 00  --private-key 0x850683b40d4a740aa6e745f889a6fdc8327be76e122f5aba645a5b02d0248db8
```

The expected output should be:
```
....
Call SUCCESS
=== Console Logs:
0
0
0
0
0
0
0
0
0
0
0
=== Call traces:
....
```

If you see a non-zero value, it means that one of the resulting limbs does not equal the expected value.

**Review by jpcenteno — APPROVED** [review]

File: precompiles/Modexp.yul [hunk 4]
```diff
@@ -333,6 +333,54 @@ object "ModExp" {
...
+                let limbOffset := 0
+                for {let i := numberOfLimbs} gt(i, 0) {i := sub(i, 1)} {
+                    limbOffset := mul(sub(i,1), 32)
+                    leftIthLimbValue := getLimbValueAtOffset(lhsPointer, limbOffset)
+                    rightIthLimbValue := getLimbValueAtOffset(rhsPointer, limbOffset)
+                    ithLimbSubtractionResult, borrow :=
+                                               subLimbsWithBorrow(leftIthLimbValue, rightIthLimbValue, borrow)
+                    storeLimbValueAtOffset(resultPointer, limbOffset, ithLimbSubtractionResult)
+
+                }
```
Comment (ilitteri) [inline precompiles/Modexp.yul:381]: We can get rid of `numberOfLimbs - 1` `mul` operations if we start `limbOffset` at `shl(5, numberOfLimbs)` and adjust it subtracting LIMB_SIZE_IN_BYTES() at the end of every iteration.
```suggestion
                let limbOffset := shl(5, numberOfLimbs)
                for {let i := numberOfLimbs} gt(i, 0) {i := sub(i, 1)} {
                    leftIthLimbValue := getLimbValueAtOffset(lhsPointer, limbOffset)
                    rightIthLimbValue := getLimbValueAtOffset(rhsPointer, limbOffset)
                    ithLimbSubtractionResult, borrow :=
                                               subLimbsWithBorrow(leftIthLimbValue, rightIthLimbValue, borrow)
                    storeLimbValueAtOffset(resultPointer, limbOffset, ithLimbSubtractionResult)
                    limbOffset := sub(limbOffset, LIMB_SIZE_IN_BYTES())
                }
```

**Review by ilitteri — COMMENTED** [review]
Nice work! Everything looks good, I suggested a little optimization for the main loop.

**Review by ilitteri — APPROVED** [review]

File: precompiles/Modexp.yul (hunk 4 above)
Comment (fkrause98) [reply precompiles/Modexp.yul:381]: Sure, I'll test this again after the suggested changes and let you know.

Comment (fkrause98) [reply precompiles/Modexp.yul]: These changes make the algorithm give the correct output, I'll ignore them for now. Are they really needed? @ilitteri

**Comment by jpcenteno** [comment]
closes #133

**Pull request merged**